# EP213: Context-Aware String Concatenation

**Book Reference**: "Effective Python" (3rd Edition), Item 13, Chapter 2: Lists and Dictionaries
**Tier**: 1 (High Impact)
**Status**: Tier 1 Priority

## Overview

Prefer explicit string concatenation over implicit concatenation, especially in collections where missing commas can cause silent bugs. Context matters - implicit concatenation is dangerous in lists/tuples but acceptable in some other contexts.

## Problem

Implicit string concatenation in collections can hide bugs:

```python
# ❌ Missing comma - creates one string instead of two!
items = [
    "first_item" "second_item",  # BUG: Missing comma!
    "third_item",
]
# Result: ["first_itemsecond_item", "third_item"] - NOT what we wanted!

# ❌ More subtle version
config = [
    "database_host" "database_port",  # Missing comma - silent bug
    "redis_url",
    "api_key",
]

# ❌ In function calls
process_items([
    "item1" "item2",  # Missing comma creates unexpected behavior
    "item3"
])
```

## Solution

Use explicit concatenation or fix the missing comma:

```python
# ✅ Fix the missing comma (usually the right solution)
items = [
    "first_item",     # Comma added
    "second_item",    # Clear separation
    "third_item",
]

# ✅ Explicit concatenation when intentional
items = [
    "first_item" + "second_item",  # Clearly intentional
    "third_item",
]

# ✅ Even better - use f-strings for readability
prefix = "first"
suffix = "second"
items = [
    f"{prefix}_item{suffix}_item",  # Clear intent
    "third_item",
]
```

## Detection Criteria

This rule detects:

1. **Implicit concatenation in lists/tuples**:
   ```python
   items = ["string1" "string2", "other"]  # Dangerous
   ```

2. **Implicit concatenation in function arguments**:
   ```python
   func("arg1" "arg2", other_arg)  # Likely missing comma
   ```

3. **Multi-line implicit concatenation in collections**:
   ```python
   data = [
       "line1"
       "line2",  # Missing comma between lines
       "line3"
   ]
   ```

## When Not to Apply

- **Intentional concatenation outside collections**:
  ```python
  # ✅ Acceptable in simple contexts
  message = "Hello " "World"  # Clear intent, not in collection
  ```

- **String literals for readability**:
  ```python
  # ✅ Long strings broken for line length
  long_string = ("This is a very long string that "
                 "needs to be broken across lines")
  ```

- **Template strings**:
  ```python
  # ✅ SQL or similar templates
  query = ("SELECT * FROM users "
           "WHERE active = true")
  ```

## Book Context

From Item 13: "Adjacent string literals are concatenated together at compile time... This can introduce subtle bugs when you forget commas between strings in a list."

The book emphasizes that:
- Implicit concatenation can hide missing commas
- Context determines whether it's dangerous
- Collections are the most error-prone context
- Explicit is better than implicit (PEP 20)

## Gap Analysis

**Why existing tools miss this**:
- `flake8-implicit-str-concat` (ISC001-003) detects implicit concatenation but lacks context awareness
- It flags ALL implicit concatenation, creating noise
- Our rule focuses on dangerous contexts (collections, function args)

## Implementation Notes

**AST Pattern Detection**:
1. Visit `ast.List`, `ast.Tuple`, and `ast.Call` nodes
2. Check for adjacent string literals without commas
3. Analyze parent context to determine danger level
4. Flag cases where missing comma is likely

**Context Analysis**:
- High danger: Lists, tuples, function arguments
- Medium danger: Dictionary values, set literals
- Low danger: Simple assignment, return statements

## Examples from the Wild

```python
# ❌ Real-world bug pattern
ERROR_MESSAGES = [
    "Invalid username" "or password",  # Missing comma!
    "Database connection failed",
    "Permission denied",
]
# User sees: "Invalid usernameor password" - confusing!

# ❌ Configuration lists
ALLOWED_HOSTS = [
    "localhost" "127.0.0.1",  # Missing comma creates invalid host
    "production.example.com",
]

# ❌ API endpoint definitions
ENDPOINTS = [
    "/api/users" "/profile",  # Wrong! Should be two endpoints
    "/api/settings",
]

# ✅ Fixed versions
ERROR_MESSAGES = [
    "Invalid username or password",  # Fixed - one message
    # OR
    "Invalid username",              # Two separate messages
    "Invalid password",
    "Database connection failed",
    "Permission denied",
]

ALLOWED_HOSTS = [
    "localhost",
    "127.0.0.1",                    # Separate hosts
    "production.example.com",
]

ENDPOINTS = [
    "/api/users",                   # Two endpoints
    "/api/profile",
    "/api/settings",
]
```

## Related Rules

- **EP215**: Avoid Striding and Slicing Together (Tier 3) - string complexity
- **HP001**: String concatenation in loops (High Performance Python)

## Error Message Format

```
EP213 Implicit string concatenation in collection, missing comma?
→ 'Effective Python' (3rd Edition), Item 13, Chapter 2: Lists and Dictionaries
→ Bug Prevention: Missing commas create silent errors in collections
→ Example: ["item1" "item2", "item3"] → ["item1", "item2", "item3"]
```

## Configuration Options

```ini
[flake8]
# Focus only on dangerous contexts (recommended)
ep213-strict-mode = false

# Flag all implicit concatenation (noisy but thorough)
ep213-strict-mode = true
```
