# EP320: Loop Variables After Loop Ends

**Book Reference**: "Effective Python" (3rd Edition), Item 20, Chapter 3: Functions
**Tier**: 1 (High Impact)
**Status**: Tier 1 Priority

## Overview

Never use for loop variables after the loop ends. Loop variables persist after the loop completes, but their values may be unexpected or undefined if the loop never ran or broke early. This leads to subtle bugs that are hard to debug.

## Problem

Loop variables persist with potentially unexpected values:

```python
# ❌ Using loop variable after loop - dangerous!
def find_admin_user(users):
    for user in users:
        if user.is_admin:
            break

    # BUG: What if no admin was found? What if users is empty?
    if user.is_admin:           # 'user' may be undefined!
        grant_access(user)      # This could crash or use wrong user
    else:
        deny_access()

# ❌ More subtle version
def process_items(items):
    for item in items:
        if item.should_process():
            result = process(item)

    # BUG: 'item' and 'result' persist but may not be what you expect
    log_final_item(item)        # Could be wrong item or undefined
    return result               # Could be undefined!

# ❌ Empty list edge case
def find_largest(numbers):
    for num in numbers:
        if num > 100:
            largest = num
            break

    # BUG: If numbers is empty, 'num' is undefined
    print(f"Processed {num}")   # NameError if numbers is empty!
```

## Solution

Use defensive programming patterns:

```python
# ✅ Defensive approach with explicit variable
def find_admin_user(users):
    admin_user = None
    for user in users:
        if user.is_admin:
            admin_user = user
            break

    if admin_user is not None:
        grant_access(admin_user)
    else:
        deny_access()

# ✅ Even better - return early
def find_admin_user(users):
    for user in users:
        if user.is_admin:
            return user
    return None  # Explicit "not found"

# ✅ Using built-in functions
def find_admin_user(users):
    admin_users = [user for user in users if user.is_admin]
    return admin_users[0] if admin_users else None

# ✅ Using next() with default
def find_admin_user(users):
    return next((user for user in users if user.is_admin), None)

# ✅ Handle empty collections properly
def find_largest(numbers):
    largest = None
    for num in numbers:
        if num > 100:
            largest = num
            break

    if largest is not None:
        print(f"Found large number: {largest}")
    else:
        print("No large numbers found")
```

## Detection Criteria

This rule detects:

1. **Direct loop variable usage after loop**:
   ```python
   for item in items:
       process(item)

   result = item.value  # Using 'item' after loop - triggers EP320
   ```

2. **Loop variable in conditional after loop**:
   ```python
   for user in users:
       if user.is_admin:
           break

   if user.is_admin:    # Dangerous usage - triggers EP320
       grant_access(user)
   ```

3. **Loop variable in function calls after loop**:
   ```python
   for data in dataset:
       if data.is_valid():
           break

   log_data(data)       # 'data' may be undefined - triggers EP320
   ```

## When Not to Apply

- **Intentional final value usage**:
  ```python
  # ✅ When you genuinely want the last value
  for i in range(10):
      pass

  print(f"Final counter: {i}")  # i=9, this is intentional
  ```

- **Accumulator patterns**:
  ```python
  # ✅ Building up a result
  total = 0
  for value in values:
      total += value

  return total  # 'total' accumulates throughout loop
  ```

- **Loop variable used immediately in same statement**:
  ```python
  # ✅ Same line usage (less risky)
  for item in items:
      if should_stop(item):
          break
  else:
      item = default_item  # for-else ensures 'item' is defined

  process(item)  # Safe due to else clause
  ```

## Book Context

From Item 20: "Using the loop variable after the loop has ended can lead to surprising results. The value of the loop variable is left over from the last iteration of the loop."

Key insights from the book:
- Loop variables persist in Python (unlike some other languages)
- Values can be from last iteration, broken iteration, or undefined
- Empty sequences make this especially dangerous
- Defensive programming prevents these bugs

## Gap Analysis

**Why this is different from flake8-bugbear B023**:
- **B023**: Focuses on closure/lambda late-binding issues
- **EP320**: Focuses on direct post-loop variable usage
- **Different patterns**: B023 = `[lambda: x for x in items]`, EP320 = `for x in items: ...; use(x)`
- **No overlap**: Completely different bug categories

## Implementation Notes

**AST Pattern Detection**:
1. Track for-loop variable names in each scope
2. Scan subsequent statements in same scope
3. Flag usage of loop variables after loop ends
4. Handle nested loops and scope boundaries

**Scope Analysis**:
- Function scope: Most important to check
- Module scope: Less critical but still problematic
- Class scope: Methods and properties
- Nested loops: Inner loop variables vs outer loop variables

**Edge Cases**:
- `for-else` constructs (safer usage)
- Loop variable reassignment after loop
- Same variable name reused in different loops

## Examples from the Wild

```python
# ❌ Authentication bug
def authenticate_user(credentials_list):
    for creds in credentials_list:
        if validate(creds):
            break

    # BUG: If validation fails for all, 'creds' has last invalid value
    log_authentication(creds.username)  # Wrong user logged!

# ❌ File processing bug
def process_files(file_paths):
    for path in file_paths:
        if path.endswith('.txt'):
            content = read_file(path)
            break

    # BUG: If no .txt files found, 'path' and 'content' are undefined
    backup_file(path)     # Could be wrong file or NameError
    return content        # Could be undefined

# ❌ Database query bug
def find_user_by_email(users, target_email):
    for user in users:
        if user.email == target_email:
            break

    # BUG: If email not found, 'user' has last user from list
    return user  # Wrong user returned!

# ✅ Fixed versions
def authenticate_user(credentials_list):
    authenticated_user = None
    for creds in credentials_list:
        if validate(creds):
            authenticated_user = creds
            break

    if authenticated_user:
        log_authentication(authenticated_user.username)
        return authenticated_user
    else:
        log_authentication_failure()
        return None

def process_files(file_paths):
    for path in file_paths:
        if path.endswith('.txt'):
            return read_file(path)  # Return immediately

    return None  # Explicit "not found"

def find_user_by_email(users, target_email):
    return next((user for user in users if user.email == target_email), None)
```

## Advanced Patterns

```python
# ✅ For-else pattern (safer)
def find_item(items, condition):
    for item in items:
        if condition(item):
            break
    else:
        # This runs only if loop completed without break
        item = None  # Explicit assignment

    return item  # Safe to use

# ✅ Walrus operator pattern (Python 3.8+)
def process_until_condition(items):
    for item in items:
        if (result := process(item)) is not None:
            break
    else:
        result = None

    return result

# ✅ Iterator pattern
def find_first_match(items, predicate):
    iterator = (item for item in items if predicate(item))
    return next(iterator, None)  # Clean and safe
```

## Related Rules

- **EP321**: Be Defensive when Iterating over Arguments (Tier 1) - related iterator safety
- **EP531**: Return Objects vs >3 Tuple Unpacking (Tier 2) - function return patterns

## Error Message Format

```
EP320 Loop variable used after loop ends, use defensive assignment
→ 'Effective Python' (3rd Edition), Item 20, Chapter 3: Functions
→ Bug Prevention: Loop variables may be undefined or have unexpected values
→ Example: for x in items: ...; use(x) → result = None; for x in items: result = x; use(result)
```

## Performance Notes

- Defensive patterns have minimal performance overhead
- Built-in functions (`next`, comprehensions) are often faster
- Early return patterns eliminate unnecessary iterations
- Generator expressions are memory-efficient for large datasets
